pseudoanonymize/base.py

```diff
@@ -0,0 +1,73 @@
+import logging
+import os
+from abc import ABC, abstractmethod
+from typing import Any, Dict, Optional
+
+from faker import Faker
+
+from .exceptions import MaxRetriesExceededException, UnparsableLLMOutputException
+
+fake = Faker()
+
+
+class BaseProcessor(ABC):
+    def __init__(self, client: Any):
+        self.client = client
+        self.logger = self._setup_logger()
+
+    @abstractmethod
+    def predict(self, input_: Dict[str, Any]) -> Dict[str, Any]:
+        pass
+
+    def retry_prediction(self, input_: Dict[str, Any], max_retries: int = 5) -> Optional[Dict[str, Any]]:
+        retries = 0
+        while retries < max_retries:
+            try:
+                prediction = self.predict(input_)
+                return prediction
+            except UnparsableLLMOutputException as e:
+                retries += 1
+                if retries >= max_retries:
+                    self.logger.error(f"Failed to process item after {max_retries} attempts")
+                    raise MaxRetriesExceededException(f"Failed to process item after {max_retries} attempts")
+
+    def _setup_logger(self) -> logging.Logger:
+        logger = logging.getLogger(self.__class__.__name__)
+        logger.setLevel(logging.DEBUG)
+        log_dir = 'logs'
+        os.makedirs(log_dir, exist_ok=True)
+        handler = logging.FileHandler(os.path.join(log_dir, f'{self.__class__.__name__}.log'))
+        handler.setLevel(logging.DEBUG)
+        formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
+        handler.setFormatter(formatter)
+        logger.addHandler(handler)
+        return logger
```

Comment:
What do we intend to log to file when the service is running?

Reply:
My intention was to mainly log the outputs of the model that caused an UnparsableLLMOutputException.